src/data/int/order.lean

```diff
@@ -0,0 +1,73 @@
+/-
+Copyright (c) 2021 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import data.int.basic
+import order.conditionally_complete_lattice
+/-!
+## `ℤ` forms a conditionally complete linear order
+
+The integers form a conditionally complete linear order.
+-/
+
+open int
+open_locale classical
+noncomputable theory
+
+instance : conditionally_complete_linear_order ℤ :=
+{ Sup := λ s, if h : s.nonempty ∧ bdd_above s then
+    (greatest_of_bdd (classical.some h.2) (classical.some_spec h.2) h.1).1 else 0,
+  Inf := λ s, if h : s.nonempty ∧ bdd_below s then
+    (least_of_bdd (classical.some h.2) (classical.some_spec h.2) h.1).1 else 0,
+  le_cSup := begin
+    intros s n hs hns,
+    have : s.nonempty ∧ bdd_above s := ⟨⟨n, hns⟩, hs⟩,
+    rw [dif_pos this],
+    exact (greatest_of_bdd _ _ _).2.2 n hns
+  end,
+  cSup_le := begin
+    intros s n hs hns,
+    have : s.nonempty ∧ bdd_above s := ⟨hs, ⟨n, hns⟩⟩,
+    rw [dif_pos this],
+    exact hns (greatest_of_bdd _ (classical.some_spec this.2) _).2.1
+  end,
+  cInf_le := begin
+    intros s n hs hns,
+    have : s.nonempty ∧ bdd_below s := ⟨⟨n, hns⟩, hs⟩,
+    rw [dif_pos this],
+    exact (least_of_bdd _ _ _).2.2 n hns
+  end,
+  le_cInf := begin
+    intros s n hs hns,
+    have : s.nonempty ∧ bdd_below s := ⟨hs, ⟨n, hns⟩⟩,
+    rw [dif_pos this],
+    exact hns (least_of_bdd _ (classical.some_spec this.2) _).2.1
+  end,
+  .. int.linear_order, ..lattice_of_linear_order }
+
+namespace int
+
+@[simp]
+lemma cSup_empty : Sup (∅ : set ℤ) = 0 := dif_neg (by simp)
+
+lemma cSup_of_not_bdd_above {s : set ℤ} (h : ¬ bdd_above s) : Sup s = 0 := dif_neg (by simp [h])
```

Comment:
I think a lemma that says something like
```lean
lemma cSup_eq_greatest_of_bdd {s : set ℤ} [decidable_pred (∈ s)]
  (b : ℤ) (Hb : ∀ z ∈ s → z ≤ b) (Hinh : ∃ z : ℤ, z ∈ s) :
  Sup s = greatest_of_bdd b Hb Hinh := dif_pos sorry
```
would be good - notably, that allows rewriting `greatest_of_bdd` into `cSup` so that various `conditionally_complete_lattice` lemmas can be applied